EIPS/eip-7928.md

```diff
@@ -0,0 +1,199 @@
+
+### State Transition Function
+
+Modify the state transition function to validate the block-level access lists:
+
+```python 
+def state_transition(block: Block) -> None:
+    computed_access_list = {}
+    computed_balance_diffs = {}
+    computed_nonce_diffs = {}
+
+    for idx, tx in enumerate(block.transactions):
+        # Execute transaction and track state accesses
+        accessed_items, balance_changes = execute_transaction(tx)
+        
+        # Process storage accesses
+        for (address, slot), is_write, value in accessed_items:
+            if (address, slot) not in computed_access_list:
+                computed_access_list[(address, slot)] = []
+            
+            access_entry = {"tx_index": idx}
+            if is_write:
+                access_entry["value_after"] = value
+            
+            computed_access_list[(address, slot)].append(access_entry)
+        
+        # Process balance changes
+        for address, delta in balance_changes:
+            if address not in computed_balance_diffs:
+                computed_balance_diffs[address] = []
+            computed_balance_diffs[address].append({"tx_index": idx, "delta": delta})
+
+    # Validate access list and balance diffs
+    if not validate_access_list(block.block_access_list, computed_access_list):
+        raise InvalidBlock("Mismatch in block-level access list.")
+        
+    if not validate_balance_diffs(block.balance_diffs, computed_balance_diffs):
+        raise InvalidBlock("Mismatch in balance diffs.")
+        
+    if not validate_nonce_diffs(block.nonce_diffs, computed_nonce_diffs):
+        raise InvalidBlock("Mismatch in nonce diffs.")
+```
+
+The BAL MUST be complete and accurate. It MUST NOT contain too few entries (missing accesses) or too many entries (spurious accesses). Any missing or extra entries in the access list, balance diffs, or nonce diffs SHALL result in block invalidation.
+
+Client implementations MUST compare the accessed addresses and storage keys gathered during execution (as defined in [EIP-2929](./eip-2929.md)) with those included in the BAL to determine validity.
+
+Client implementations MAY invalidate the block right away in case any transaction steps outside the declared state.
+
+## Rationale
+
+### BAL Design Choice
+
+We chose the BAL variant for several key reasons:
+
+1. **Balance between size and parallelization benefits**: BALs enable both parallel disk reads and parallel EVM execution while keeping the size manageable.
+
+2. **Storage value inclusion**: Including post-execution values for write operations facilitates state reconstruction during syncing, enabling faster chain catch-up.
+
+3. **Balance diffs**: Including balance diffs ensures correct handling of execution that depends on account balances, crucial for parallel transaction execution.
+
+4. **Nonce diffs**: Providing pre-block nonce values for transaction senders enables validators to verify nonce ordering without sequential execution.
```

Comment:
Ok, maybe I got the nonce part here wrong. Note that for transaction 0 in the block, which has nonce N in the transction and is sent from A, we should fetch from disk account A and verify that it's nonce is N. We now also know that after this tx it is N+1. There is (currently) no way to edit the nonce during the tx (but with 7702 there will be, one could call into an account and then CREATE/CREATE2 there bumping the nonce). So nonce updates should be very predictable.

The ways to update a nonce:

- Send a transaction
- Execute a CREATE/CREATE2 operation
- Change the delegation of an account via 7702 authority lists

I think only (2) is unpredictable. The remainder can be pre-fetched and checked from disk.

Reply:
Right, this is an artifact from an older version (where pre-executio values were considered). This point just needs to be removed.